src/solvers/NelderMead.jl

```diff
@@ -80,62 +227,75 @@ function NelderMead!(
+    ξ = inverse_retract(M, m, s.population.pts[last(ind)], s.inverse_retraction_method)
     # reflect last
-    xr = retract(p.M, m, -o.α * ξ, o.retraction_method)
-    Costr = get_cost(p, xr)
+    xr = retract(M, m, -s.α * ξ, s.retraction_method)
+    Costr = get_cost(mp, xr)
+    continue_steps = true
     # is it better than the worst but not better than the best?
-    if Costr >= o.costs[first(ind)] && Costr < o.costs[last(ind)]
+    if Costr >= s.costs[first(ind)] && Costr < s.costs[last(ind)]
         # store as last
-        o.population[last(ind)] = xr
-        o.costs[last(ind)] = Costr
+        s.population.pts[last(ind)] = xr
+        s.costs[last(ind)] = Costr
+        continue_steps = false
     end
     # --- Expansion ---
-    if Costr < o.costs[first(ind)] # reflected is better than fist -> expand
-        xe = retract(p.M, m, -o.γ * o.α * ξ, o.retraction_method)
-        Coste = get_cost(p, xe)
+    if Costr < s.costs[first(ind)] # reflected is better than fist -> expand
+        xe = retract(M, m, -s.γ * s.α * ξ, s.retraction_method)
+        Coste = get_cost(mp, xe)
         # successful? use the expanded, otherwise still use xr
-        o.population[last(ind)] .= Coste < Costr ? xe : xr
-        o.costs[last(ind)] = min(Coste, Costr)
+        s.population.pts[last(ind)] .= Coste < Costr ? xe : xr
+        s.costs[last(ind)] = min(Coste, Costr)
+        continue_steps = false
     end
     # --- Contraction ---
-    if Costr > o.costs[ind[end - 1]] # even worse than second worst
-        s = (Costr < o.costs[last(ind)] ? -o.ρ : o.ρ)
-        xc = retract(p.M, m, s * ξ, o.retraction_method)
-        Costc = get_cost(p, xc)
-        if Costc < o.costs[last(ind)] # better than last ? -> store
-            o.population[last(ind)] = xc
-            o.costs[last(ind)] = Costc
+    if continue_steps && Costr > s.costs[ind[end - 1]] # even worse than second worst
+        step = (Costr < s.costs[last(ind)] ? -s.ρ : s.ρ)
+        xc = retract(M, m, step * ξ, s.retraction_method)
+        Costc = get_cost(mp, xc)
+        if Costc < s.costs[last(ind)] # better than last ? -> store
+            s.population.pts[last(ind)] = xc
+            s.costs[last(ind)] = Costc
+            continue_steps = false
         end
     end
     # --- Shrink ---
-    for i in 2:length(ind)
-        o.population[ind[i]] = retract(
-            p.M,
-            o.population[ind[1]],
-            inverse_retract(
-                p.M, o.population[ind[1]], o.population[ind[i]], o.inverse_retraction_method
-            ),
-            o.σ,
-            o.retraction_method,
-        )
-        # update cost
-        o.costs[ind[i]] = get_cost(p, o.population[ind[i]])
+    if continue_steps
+        for i in 2:length(ind)
+            retract!(
+                M,
+                s.population.pts[ind[i]],
+                s.population.pts[ind[1]],
+                inverse_retract(
```

Comment:
I am not sure whether we lose in number of iterations or “just” in time. For time we could allocate one tangent vector (maybe even a temporary vector in the state) where we do this inverse retract inplace  (then in the beginning of the loop, that is before 281), so that we have an allocation less here.
For the rest we might need a large Cache for the cost at least.

Reply:
Compared to Optim, our Nelder-Mead uses orders of magnitude more objective evaluations and still reaches a worse minimum so optimizing allocation is definitely not the biggest problem here.

Reply:
Well, without caching we are reevaluating the cost several times for each point I fear, but that would only be solved by the complicated cache.

Reply:
ah but as a simple solution you could use `@memoize` for your cost in your example that should help already.

Reply:
Our Nelder-Mead had another bug. After fixing it, the convergence it actually similar to Optim.jl.

Reply:
Bug? I do not do bugs, I always(!) implement without any errors 😬 – nice that you found it!

Reply:
Looking forward to what that was ;)

Reply:
:smile:

Reply:
The issue was that the mean calculated at the beginning should not include the worst point in simplex :slightly_smiling_face: .

Reply:
Indeed, Dreisigmeyer writes that as well, see https://optimization-online.org/2007/08/1742/ page 4, where he excludes p_{n+1}, from the mean \bar p; I must have missed that.

Reply:
I've just fixed it, and also #183, because it was bugging me.